crates/moss-jsonlogic/src/rule_with_validation.rs

```diff
@@ -0,0 +1,1171 @@
+use crate::rule::{Operator, Rule};
+use serde::{Deserialize, Serialize, Serializer};
+use serde_json::Value;
+use std::fmt::{Debug, Display};
+use std::ops::Not;
+use std::ops::{Add, BitAnd, BitOr, Div, Mul, Sub};
+use thiserror::Error;
+
+// TODO: update documentation
+
+#[derive(Debug, Error)]
+pub enum RuleError {
+    #[error("Operand '{operand:?}' have invalid type for operator '{operator}'")]
+    InvalidType { operator: Operator, operand: Rule },
+    #[error(
+        "Operand '{left:?}' has incompatible type with operand '{right:?}' for equality checks"
+    )]
+    IncompatibleType { left: Rule, right: Rule },
+    #[error("Cannot divide by zero")]
+    ZeroDivision,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+enum ResultType {
+    Number,
+    String,
+    Boolean,
+    Array,
+    Object,
+    Variable,
+    Undefined,
+}
+
+trait RuleType {
+    fn get_type(&self) -> ResultType;
+}
+
+impl RuleType for Operator {
+    fn get_type(&self) -> ResultType {
+        match self {
+            Operator::Equal => ResultType::Boolean,
+            Operator::NotEqual => ResultType::Boolean,
+            Operator::GreaterThan => ResultType::Boolean,
+            Operator::LessThan => ResultType::Boolean,
+            Operator::GreaterThanOrEqual => ResultType::Boolean,
+            Operator::LessThanOrEqual => ResultType::Boolean,
+            Operator::And => ResultType::Boolean,
+            Operator::Or => ResultType::Boolean,
+            Operator::Not => ResultType::Boolean,
+            Operator::Add => ResultType::Number,
+            Operator::Subtract => ResultType::Number,
+            Operator::Multiply => ResultType::Number,
+            Operator::Divide => ResultType::Number,
+            Operator::Modulo => ResultType::Number,
+            Operator::In => ResultType::Boolean,
+            Operator::Cat => ResultType::Undefined,
+            Operator::Map => ResultType::Variable,
+            Operator::Reduce => ResultType::Variable,
+            Operator::Filter => ResultType::Array,
+            Operator::All => ResultType::Boolean,
+            Operator::None => ResultType::Boolean,
+            Operator::Some => ResultType::Boolean,
+            Operator::Merge => ResultType::Array,
+            Operator::If => ResultType::Undefined,
+            Operator::Var => ResultType::Variable,
+            Operator::Missing => ResultType::Undefined,
+            Operator::MissingSome => ResultType::Undefined,
+        }
+    }
+}
+
+pub struct RuleWithValidation {
+    rule: Rule,
+}
```

Comment:
Let’s name the type without validation `RawRule` and the validated type `ValidatedRule`